File: Middle/src/it/dani/selfhome/way/WhereAreYouListener.java
package it.dani.selfhome.way;

import org.jetbrains.annotations.NotNull;

import java.io.*;
import java.net.*;
import java.util.Arrays;
import java.util.Enumeration;

public class WhereAreYouListener extends Thread {

    private final InetAddress address;
    private final int listenPort;
    private final int servicePort;

    public WhereAreYouListener(@NotNull InetAddress address, int listenPort, int servicePort)
    {
        this.address = address;
        this.listenPort = listenPort;
        this.servicePort = servicePort;
    }

    @Override
    public void run()
    {
        try(MulticastSocket sock = new MulticastSocket(this.listenPort))
        {
            sock.setReuseAddress(true);
            sock.setSoTimeout(0);
            sock.joinGroup(this.address);
            byte[] buffer = new byte[2048];
            DatagramPacket pack = new DatagramPacket(buffer, buffer.length);

            while(true)
            {
                try
                {
                    Arrays.fill(buffer,(byte)0);
                    pack.setData(buffer);
                    sock.receive(pack);
                    System.out.println("Received (middle): WAY from " + pack.getAddress().getHostAddress());

                    String response = this.handleRequest(pack);

                    ByteArrayOutputStream baos = new ByteArrayOutputStream(128);
                    PrintWriter pw = new PrintWriter(baos);
                    pw.print(response + "\0");
                    pw.flush();
                    pack.setData(baos.toByteArray());
                    pw.close();
                    sock.send(pack);

                } catch (IOException e) {
                    e.printStackTrace();
                }
            }

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private String handleRequest(DatagramPacket pack) throws UnknownHostException, SocketException {
        Enumeration<NetworkInterface> interfaces = NetworkInterface.getNetworkInterfaces();

        String result = null;

        while(result == null && interfaces.hasMoreElements())
        {
            NetworkInterface networkInterface = interfaces.nextElement();

            for(InterfaceAddress ia : networkInterface.getInterfaceAddresses())
            {
                if(result == null && isInSameNet(pack.getAddress(),ia))
                {
                    result = ia.getAddress().getHostAddress();
                }
            }
        }

        return ((result == null)? InetAddress.getLocalHost().getHostAddress() : result) + ":" + this.servicePort;
    }

    private boolean isInSameNet(InetAddress addr, InterfaceAddress interfaceAddress)
    {
        boolean result = true;

        for(int count = 0; result && count < addr.getAddress().length && count < interfaceAddress.getNetworkPrefixLength()/8; count++)
        {
            result = (addr.getAddress()[count] == interfaceAddress.getAddress().getAddress()[count]);
        }

        return result;
    }
}

File: Middle/src/main/kotlin/it/dani/way/WhereAreYouListener.kt
package it.dani.way

import it.dani.listener.Listener
import java.io.IOException
import java.net.DatagramPacket
import java.net.InetAddress
import java.net.InterfaceAddress
import java.net.MulticastSocket
import java.net.NetworkInterface

class WhereAreYouListener(address: InetAddress, listenPort: Int, servicePort: Int): Listener {
    private val thread = Thread {
        with(MulticastSocket(listenPort)) {
            this.reuseAddress = true
            this.soTimeout = 0
            this.joinGroup(address)

            while(true) {
                val buffer = ByteArray(512)
                val pack = DatagramPacket(buffer, buffer.size)
                try {
                    this.receive(pack)

                    Thread {
                        println("Received: WAY from ${pack.address.hostAddress}")
                        val response = this@WhereAreYouListener.handleRequest(pack, servicePort)
                        println("response: $response")
                        pack.data = response.toByteArray() + 0
                        this.send(pack)
                    }.start()

                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        }
    }

    override fun start() {
        this.thread.start()
    }

    override fun stop() {
        this.thread.interrupt()
    }

    private fun handleRequest(packet: DatagramPacket, servicePort: Int): String {
        var result: String? = null

        NetworkInterface.getNetworkInterfaces().iterator().forEach { networkInterface ->
            networkInterface.interfaceAddresses.forEach { interfaceAddress ->
                if(result == null && this.isInSameNet(packet.address, interfaceAddress)) {
                    result = interfaceAddress.address.hostAddress
                }
            }
        }
        return "${result ?: InetAddress.getLocalHost().hostAddress}:$servicePort"
    }

    private fun isInSameNet(address: InetAddress, interfaceAddress: InterfaceAddress): Boolean {
        var result = address.address.size == interfaceAddress.address.address.size

        if(result) {
            for(count in 0..<(interfaceAddress.networkPrefixLength/8)) {
                if(address.address[count] != interfaceAddress.address.address[count]) {
                    result = false
                }
            }
        }

        return result
    }
}